src/collections/map.rs

```diff
@@ -562,7 +712,650 @@ mod tests {
+        match iter_next {
+            Some((key, value)) => {
+                assert_eq!(key, 14);
+                assert_eq!(*value, 28);
+            }
+            None => assert!(false),
+        }
+
+        let iter_next = Map::iter_merge_next(&mut map_iter, &mut range_iter).unwrap();
+        match iter_next {
+            Some(_) => assert!(false),
+            None => (),
+        }
+    }
+
+    #[test]
+    fn iter_merge_next_store_only() {
+        let store = Store::new(MapStore::new());
+        let mut edit_map: Map<u32, u32> = Map::create(store.clone(), ()).unwrap();
+
+        edit_map.entry(12).unwrap().or_insert(24).unwrap();
+        edit_map.entry(13).unwrap().or_insert(26).unwrap();
+        edit_map.entry(14).unwrap().or_insert(28).unwrap();
+
+        edit_map.flush().unwrap();
+
+        let read_map: Map<u32, u32> = Map::create(store.clone(), ()).unwrap();
+
+        let mut map_iter = read_map.children.range(..).peekable();
+        let mut range_iter = read_map.store.range(..).peekable();
+
+        let iter_next = Map::iter_merge_next(&mut map_iter, &mut range_iter).unwrap();
+        match iter_next {
+            Some((key, value)) => {
+                assert_eq!(key, 12);
+                assert_eq!(*value, 24);
+            }
+            None => assert!(false),
+        }
+
+        let iter_next = Map::iter_merge_next(&mut map_iter, &mut range_iter).unwrap();
+        match iter_next {
+            Some((key, value)) => {
+                assert_eq!(key, 13);
+                assert_eq!(*value, 26);
+            }
+            None => assert!(false),
+        }
+
+        let iter_next = Map::iter_merge_next(&mut map_iter, &mut range_iter).unwrap();
+        match iter_next {
+            Some((key, value)) => {
+                assert_eq!(key, 14);
+                assert_eq!(*value, 28);
+            }
+            None => assert!(false),
+        }
+
+        let iter_next = Map::iter_merge_next(&mut map_iter, &mut range_iter).unwrap();
+        match iter_next {
+            Some(_) => assert!(false),
+            None => (),
+        }
+    }
+
+    #[test]
+    fn iter_merge_next_mem_remove() {
+        let store = Store::new(MapStore::new());
+        let mut edit_map: Map<u32, u32> = Map::create(store.clone(), ()).unwrap();
+
+        edit_map.entry(12).unwrap().or_insert(24).unwrap();
+
+        edit_map.flush().unwrap();
+
+        let mut read_map: Map<u32, u32> = Map::create(store.clone(), ()).unwrap();
+
+        read_map.remove(12).unwrap();
+
+        let mut map_iter = read_map.children.range(..).peekable();
+        let mut range_iter = read_map.store.range(..).peekable();
+
+        let iter_next = Map::iter_merge_next(&mut map_iter, &mut range_iter).unwrap();
+        match iter_next {
+            Some((key, value)) => {
+                assert_eq!(key, 12);
+                assert_eq!(*value, 24);
+            }
+            None => assert!(false),
+        }
```

Comment:
I would expect the removal to apply to the values emitted by the iterator, the test should ensure the entry isn't emitted.